build/build.c

```diff
@@ -196,9 +198,109 @@ rpmRC doScript(rpmSpec spec, rpmBuildFlags what, const char *name,
     return rc;
 }
 
+static rpmRC doBuildRequires(rpmSpec spec, int test) {
+    char * buildDir = rpmGenPath(spec->rootDir, "%{_builddir}", "");
+    char *scriptName = NULL;
+    char * buildCmd = NULL;
+    FD_t fd = NULL;
+    FILE * fp = NULL;
+    StringBuf sb_stdout = NULL;
+    int argc = 0;
+    const char **argv = NULL;
+
+    int outc;
+    ARGV_t output = NULL;
+
+    rpmRC rc = 1; /* assume failure */
+
+    if (!spec->buildrequires) {
+	rc = RPMRC_OK;
+	goto exit;
+    }
+
+    /* write script to disk */
+    fd = rpmMkTempFile(spec->rootDir, &scriptName);
+    if (Ferror(fd)) {
+	rpmlog(RPMLOG_ERR, _("Unable to open temp file: %s\n"), Fstrerror(fd));
+	goto exit;
+    }
+
+    if ((fp = fdopen(Fileno(fd), "w")) == NULL) {
+	rpmlog(RPMLOG_ERR, _("Unable to open stream: %s\n"), strerror(errno));
+	goto exit;
+    }
+
+    fprintf(fp, "cd '%s'\n", buildDir);
+
+    if (spec->buildSubdir)
+    	fprintf(fp, "cd '%s'\n", spec->buildSubdir);
+    fprintf(fp, "%s", getStringBuf(spec->buildrequires));
+    (void) fclose(fp);
```

Comment:
This script-running part with all the templating etc needs to be merged into doScript() really.
Make doScript() always use rpmfcExec() and return its output via pointer if supplied, and parse the output in the caller.

Reply:
Except that rpmfcExec() buffers the entire output so you can't watch a build live, that's not good. Maybe add a new parameter to it to optionally duplicate the output live onto a given stream.

Reply:
Something like this: https://github.com/rpm-software-management/rpm/pull/659

Reply:
Yeah, this ended up much closer to doScript() than anticipated. I started with trying to use doScript and that didn't work out. Now that most things from doScript got added merging the two functions looks much more natural.